File: src/models/indicators.py
import pandas as pd
import numpy as np
from typing import Optional
from src.utils.logger import get_logger

logger = get_logger()

class TechnicalIndicators:
    @staticmethod
    def calculate_rsi(data: pd.DataFrame, period: int = 14) -> pd.Series:
        """Calculate Relative Strength Index."""
        try:
            delta = data['close'].diff()
            gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
            loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
            rs = gain / loss
            return 100 - (100 / (1 + rs))
        except Exception as e:
            logger.error(f"Error calculating RSI: {str(e)}")
            raise
            
    @staticmethod
    def calculate_macd(
        data: pd.DataFrame,
        fast_period: int = 12,
        slow_period: int = 26,
        signal_period: int = 9
    ) -> Tuple[pd.Series, pd.Series]:
        """Calculate MACD and Signal line."""
        try:
            exp1 = data['close'].ewm(span=fast_period).mean()
            exp2 = data['close'].ewm(span=slow_period).mean()
            macd = exp1 - exp2
            signal = macd.ewm(span=signal_period).mean()
            return macd, signal
        except Exception as e:
            logger.error(f"Error calculating MACD: {str(e)}")
            raise
            
    @staticmethod
    def calculate_bollinger_bands(
        data: pd.DataFrame,
        period: int = 20,
        std_dev: int = 2
    ) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """Calculate Bollinger Bands."""
        try:
            middle_band = data['close'].rolling(window=period).mean()
            std = data['close'].rolling(window=period).std()
            upper_band = middle_band + (std * std_dev)
            lower_band = middle_band - (std * std_dev)
            return upper_band, middle_band, lower_band
        except Exception as e:
            logger.error(f"Error calculating Bollinger Bands: {str(e)}")
            raise
            
    def add_all_indicators(self, df: pd.DataFrame) -> pd.DataFrame:
        """Add all technical indicators to the dataframe."""
        try:
            # Add RSI
            df['rsi'] = self.calculate_rsi(df)
            
            # Add MACD
            df['macd'], df['macd_signal'] = self.calculate_macd(df)
            df['macd_hist'] = df['macd'] - df['macd_signal']
            
            # Add Bollinger Bands
            df['bb_upper'], df['bb_middle'], df['bb_lower'] = self.calculate_bollinger_bands(df)
            
            # Add momentum indicators
            df['momentum'] = df['close'].diff(periods=10)
            df['rate_of_change'] = df['close'].pct_change(periods=10)
            
            return df
        except Exception as e:
            logger.error(f"Error adding indicators: {str(e)}")
            raise 

File: TypeScript_version/src/ai/utils/indicators.ts
import { logger } from '../../utils/logger';

export class TechnicalIndicators {
  static calculateRSI(prices: number[], period: number = 14): number[] {
    try {
      const deltas = prices.slice(1).map((price, i) => price - prices[i]);
      const gains = deltas.map(delta => delta > 0 ? delta : 0);
      const losses = deltas.map(delta => delta < 0 ? -delta : 0);

      let avgGain = gains.slice(0, period).reduce((a, b) => a + b) / period;
      let avgLoss = losses.slice(0, period).reduce((a, b) => a + b) / period;

      const rsi = [100 - (100 / (1 + avgGain / avgLoss))];

      for (let i = period; i < prices.length - 1; i++) {
        avgGain = (avgGain * (period - 1) + gains[i]) / period;
        avgLoss = (avgLoss * (period - 1) + losses[i]) / period;
        rsi.push(100 - (100 / (1 + avgGain / avgLoss)));
      }

      return rsi;
    } catch (error) {
      logger.error('Error calculating RSI:', error);
      throw error;
    }
  }

  static calculateMACD(prices: number[], fastPeriod: number = 12, slowPeriod: number = 26, signalPeriod: number = 9): {
    macd: number[];
    signal: number[];
    histogram: number[];
  } {
    try {
      const fastEMA = this.calculateEMA(prices, fastPeriod);
      const slowEMA = this.calculateEMA(prices, slowPeriod);
      const macd = fastEMA.map((fast, i) => fast - slowEMA[i]);
      const signal = this.calculateEMA(macd, signalPeriod);
      const histogram = macd.map((value, i) => value - signal[i]);

      return { macd, signal, histogram };
    } catch (error) {
      logger.error('Error calculating MACD:', error);
      throw error;
    }
  }

  private static calculateEMA(prices: number[], period: number): number[] {
    const multiplier = 2 / (period + 1);
    const ema = [prices[0]];

    for (let i = 1; i < prices.length; i++) {
      ema.push(
        (prices[i] - ema[i - 1]) * multiplier + ema[i - 1]
      );
    }

    return ema;
  }
} 